### PR TITLE
[Feature] Support EAGLE 3

### PR DESCRIPTION
**Pull request by chromecast56**
## Motivation

Add support for EAGLE-3: https://arxiv.org/abs/2503.01840

## Modifications

- Add `EAGLE3` speculative method to server args
- Refactor `llama.py, logits_processor.py` to support capturing auxiliary hidden states
- Modify `eagle_worker.py` to support EAGLE-3 token map + untied LM head
- Add `llama_eagle3.py` model
- Tests and Documentation

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by zhyncs** [comment]
cc @Liyuhui-12 @hongyanz

**Comment by chromecast56** [comment]
Benchmarks on MT-Bench (bsz 1):

Autoregressive:

```

python -m sglang.launch_server --model-path meta-llama/Llama-3.1-8B-Instruct --port 30000 --cuda-graph-max-bs 1

#questions: 1, Throughput: 147.05 token/s, Acceptance length: 1.00
```

EAGLE-3:

```
python3 -m sglang.launch_server --model meta-llama/Llama-3.1-8B-Instruct --speculative-algo EAGLE3 \
    --speculative-draft jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B --speculative-num-steps 5 \
    --speculative-eagle-topk 8 --speculative-num-draft-tokens 64 \
    --cuda-graph-max-bs 1 --mem-fraction 0.7 --dtype float16 --port 30000

#questions: 80, Throughput: 336.61 token/s, Acceptance length: 4.29

```

**Comment by zhyncs** [comment]
@chromecast56 Can we use `--speculative-num-steps 8`

File: python/sglang/srt/models/llama.py [hunk 1]
```diff
@@ -400,10 +411,19 @@ def forward(
         input_embeds: torch.Tensor = None,
         get_embedding: bool = False,
     ) -> LogitsProcessorOutput:
-        hidden_states = self.model(input_ids, positions, forward_batch, input_embeds)
+        aux_hidden_states = None
+        if self.capture_aux_hidden_states:
```
Comment (ispobock) [inline python/sglang/srt/models/llama.py:415]: `capture_aux_hidden_states` can be initialized to False in `__init__` to fix the PR test.

File: python/sglang/srt/models/llama.py [hunk 2]
```diff
@@ -286,28 +286,39 @@ def __init__(
 
         self.norm = RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
 
+        self.capture_aux_hidden_states = False
```
Comment (ispobock) [inline python/sglang/srt/models/llama.py]: move this to `LlamaForCausalLM.__init__()`?

**Comment by Swipe4057** [comment]
@chromecast56 Could you add a comparison with the already implemented EAGLE-2?

**Comment by chromecast56** [comment]
> @chromecast56 Can we use `--speculative-num-steps 8`

Is this for the docs? Or the unit test? 8 steps is a lot of draft overhead IIRC

File: python/sglang/srt/models/llama.py (hunk 2 above)
Comment (chromecast56) [reply python/sglang/srt/models/llama.py]: `LlamaForCausalLMEagle` doesn't seem to call `super().__init__()`, so I duplicate `self.capture_aux_hidden_states = False` for now.

**Comment by chromecast56** [comment]
> @chromecast56 Could you add a comparison with the already implemented EAGLE-2?

I run EAGLE-2 on Llama-3-8B:

```
python3 -m sglang.launch_server --model meta-llama/Meta-Llama-3-8B-Instruct --speculative-algo EAGLE \
    --speculative-draft lmsys/sglang-EAGLE-LLaMA3-Instruct-8B --speculative-num-steps 5 \
    --speculative-eagle-topk 8 --speculative-num-draft-tokens 64 \
    --cuda-graph-max-bs 1 --mem-fraction 0.7 --dtype float16 --port 30000

#questions: 80, Throughput: 265.50 token/s, Acceptance length: 3.73
```

The big caveat is that Llama 3 is easier to speculate than Llama 3.1. If we want a direct comparison, we might need access to a EAGLE-2 Llama 3.1 checkpoint, or a EAGLE-3 Llama 3 checkpoint, both of which I couldn't find (cc: @Liyuhui-12 @hongyanz)

**Comment by hongyanz** [comment]
@chromecast56 Thanks for testing. Our test shows that --speculative-num-steps 8 gives the best performance for EAGLE-3. The averaged acceptance length is around 6.13, which is larger than 4.29 in your test when --speculative-num-steps 5.

We will share EAGLE-2 Llama 3.1 8B checkpoint with you.

**Comment by Liyuhui-12** [comment]
> The big caveat is that Llama 3 is easier to speculate than Llama 3.1. If we want a direct comparison, we might need access to a EAGLE-2 Llama 3.1 checkpoint, or a EAGLE-3 Llama 3 checkpoint, both of which I couldn't find (cc: @Liyuhui-12 @hongyanz)

The weights are available at this URL: https://huggingface.co/yuhuili/EAGLE-LLaMA3.1-Instruct-8B.

File: python/sglang/srt/model_executor/cuda_graph_runner.py [hunk 3]
```diff
@@ -219,7 +219,22 @@ def __init__(self, model_runner: ModelRunner):
...
+                and not model_runner.is_draft_worker
+            ):
+                self.hidden_states = torch.zeros(
+                    (
+                        self.max_num_token,
+                        3 * self.model_runner.model_config.hidden_size,
+                    ),
+                    dtype=self.model_runner.dtype,
+                )
+                # auxiliary hidden capture mode. TODO: expose this to server args?
```
Comment (rkooo567) [inline python/sglang/srt/model_executor/cuda_graph_runner.py]: +1. I was curious is 3 layers & layer_id requirement same for any model, or it depends on model size? I am also curious if it also depends on model arch (sparse vs dense) or the input type

File: python/sglang/srt/model_executor/model_runner.py [hunk 4]
```diff
@@ -197,6 +197,12 @@ def __init__(
             self.cuda_graph_runner = None
             self.init_attention_backend()
 
+        # auxiliary hidden capture mode. TODO: expose this to server args?
+        if self.spec_algorithm.is_eagle3() and not self.is_draft_worker:
+            num_layers = self.model.config.num_hidden_layers
+            layers_to_capture = [2, num_layers // 2, num_layers - 3]
```
Comment (rkooo567) [inline python/sglang/srt/model_executor/model_runner.py]: this seems hardcoded. maybe add constant or helper function?

**Review by rkooo567 — COMMENTED** [review]
LG!

**Comment by chromecast56** [comment]
@hongyanz @zhyncs I set `--speculative-num-steps 8` and compare both EAGLE + EAGLE-3 on Llama 3.1 8B. Furthermore, I set the system prompt to [this](https://github.com/SafeAILab/EAGLE/blob/main/eagle/evaluation/gen_baseline_answer_llama3chat.py#L226) (seems like the model is very sensitive to modifying the system prompt it was trained on):

```
EAGLE-3 (num_steps=8):
#questions: 80, Throughput: 378.57 token/s, Acceptance length: 5.39

EAGLE-3 (num_steps=5):
#questions: 80, Throughput: 361.70 token/s, Acceptance length: 4.68

EAGLE-2 (num_steps=8):
#questions: 80, Throughput: 230.08 token/s, Acceptance length: 3.76
```


5.36 is a bit far off from 6.13 -- not sure if this can be fully explained by environment differences. There may still be implementation issues?

**Comment by hongyanz** [comment]
@chromecast56 Hmm. Let me double check with @Liyuhui-12

Also, it is weird to see that the acceptance length 5.36 > 4.29, but the throughput 276.30 < 336.61. To me, longer acceptance length should have led to faster throughput. Did I miss anything here?

**Comment by chromecast56** [comment]
Re. throughput, I believe the overhead of drafting is more significant on SGL than the HF imp, but I havent looked into this.

**Comment by hongyanz** [comment]
Hmm. If so, won't the integration of EAGLE-3 lead to negative throughput/latency improvement if we compare Autoregressive vs. EAGLE-3?

**Comment by chromecast56** [comment]
Ah sorry, I didn't run it with cuda graphs. See updated results:
```
EAGLE-3 (num_steps=8):
#questions: 80, Throughput: 378.57 token/s, Acceptance length: 5.39

EAGLE-3 (num_steps=5):
#questions: 80, Throughput: 361.70 token/s, Acceptance length: 4.68

EAGLE-2 (num_steps=8):
#questions: 80, Throughput: 230.08 token/s, Acceptance length: 3.76

EAGLE-2 (num_steps=5):
#questions: 80, Throughput: 264.49 token/s, Acceptance length: 3.71
```

**Comment by hongyanz** [comment]
Thanks @chromecast56. Let we double check the acceptance length. It is a bit unexpected to me that when acceptance length improves from 4.68 to 5.39, the throughput improves incrementally (361.70 -> 378.57).

**Comment by chromecast56** [comment]
> Re. throughput, I believe the overhead of drafting is more significant on SGL than the HF imp, but I havent looked into this.

I believe this underlying reason still holds

File: python/sglang/srt/model_executor/cuda_graph_runner.py (hunk 3 above)
Comment (chromecast56) [reply python/sglang/srt/model_executor/cuda_graph_runner.py]: At least for EAGLE-3, seems like all the models have this requirement (3 layers, with static layer_ids) regardless of model arch + input type

**Comment by zhaochenyang20** [comment]
@PeterSH6 could you help to take a look? thansk!

**Comment by hongyanz** [comment]
@chromecast56 In the calculation of "Acceptance length", did you take into account the very last token by target model (which will always be accepted)?

And how about --speculative-num-steps?

**Comment by chromecast56** [comment]
Acceptance length is calculated as `num_output_tokens / num_verify_ct` over the entire dataset so it should take it into account. @hongyanz Do you normalize acceptance length per prompt, or is it over the entire dataset?

**Comment by hongyanz** [comment]
@chromecast56 Ours is in the round level. We calculate the number of accepted tokens in each speculative decoding round, add it by 1 (because the very last token in each round will always be accepted as it is from the target model), and average the numbers across all rounds.

**Review by ispobock — APPROVED** [review]
LGTM

**Comment by zhaochenyang20** [comment]
@chromecast56 could you fix the conflicts of the docs? @zhyncs @merrymercy  could we merge it?

**Comment by zhyncs** [comment]
@merrymercy @Ying1123 reminder

**Comment by zhyncs** [comment]
Hi @chromecast56 May you help fix the conflicts

**Review by merrymercy — APPROVED** [review]
LGTM! Thanks for the great work.

**Comment by zhaochenyang20** [comment]
cc @simveit hey simon. Eagle 3 is merged into sglang now, yienng @zhyncs will profiling it today. Could you help to update the docs https://docs.sglang.ai/backend/speculative_decoding.html after yineng provides the performance? thanks so much!

**Comment by simveit** [comment]
@zhaochenyang20 Yes. Let me read the paper in the next days.

**Comment by finger92** [comment]
should we change
"You can enable EAGLE-3 decoding by setting `--speculative_draft_model_path: EAGLE3`:"
to
"You can enable EAGLE-3 decoding by setting `--speculative-algorithm EAGLE3`:"
?

**Comment by merrymercy** [comment]
@ispobock can you share the exact commands to launch the servers for eagle2 and eagle3?
We do not need to benchmark on MT-bench, we can just use `python3 -m sglang.test.send_one`

**Comment by ispobock** [comment]
@merrymercy For bs=1, the launch command is here:
```shell
# EAGLE2
python3 -m sglang.launch_server --model meta-llama/Llama-3.1-8B-Instruct --speculative-algo EAGLE \
    --speculative-draft jamesliu1/sglang-EAGLE-Llama-3.1-Instruct-8B --speculative-num-steps 5 \
    --speculative-eagle-topk 8 --speculative-num-draft-tokens 64 \
    --cuda-graph-max-bs 1 --dtype float16 --port 30000 --tp 1 --disable-radix --mem-frac 0.7

# EAGLE3
python3 -m sglang.launch_server --model meta-llama/Llama-3.1-8B-Instruct --speculative-algo EAGLE3 \
    --speculative-draft jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B --speculative-num-steps 8 \
    --speculative-eagle-topk 8 --speculative-num-draft-tokens 64 \
    --cuda-graph-max-bs 1 --dtype float16 --port 30000 --tp 1 --disable-radix --mem-frac 0.7
```
MT-bench is the default bench dataset in [EAGLE's evaluation code](https://github.com/SafeAILab/EAGLE/blob/7a6bb3db474ef53ec6aca0a3acdba233028492d5/eagle/evaluation/gen_ea_answer_llama3chat.py#L337). It's used to keep align with the setting in the paper. 
`python3 -m sglang.test.send_one` also works for benchmark.

**Comment by ryang-max** [comment]
I'll also work on the docs about this feature these days.
cc @zhaochenyang20 @simveit
